lasagne/layers/recurrent.py

```diff
@@ -171,19 +182,23 @@ def get_output_for(self, input, mask=None, **kwargs):
         input = input.dimshuffle(1, 0, 2)
         seq_len, num_batch, _ = input.shape
 
-        # Because the input is given for all time steps, we can precompute
-        # the inputs to hidden before scanning. First we need to reshape
-        # from (seq_len, batch_size, num_inputs) to
-        # (seq_len*batch_size, num_inputs)
-        input = T.reshape(input,
-                          (seq_len*num_batch, -1))
-        input_dot_W = helper.get_output(
-            self.input_to_hidden, input, **kwargs)
-
-        # reshape to original (seq_len, batch_size, num_units)
-
-        input_dot_W = T.reshape(input_dot_W,
-                                (seq_len, num_batch, -1))
+        if self.precompute_input:
+            # Because the input is given for all time steps, we can precompute
+            # the inputs to hidden before scanning. First we need to reshape
+            # from (seq_len, batch_size, num_inputs) to
+            # (seq_len*batch_size, num_inputs)
+            input = T.reshape(input,
+                              (seq_len*num_batch, -1))
+            input_dot_W = helper.get_output(
+                self.input_to_hidden, input, **kwargs)
+
+            # reshape to original (seq_len, batch_size, num_units)
+            input_dot_W = T.reshape(input_dot_W,
+                                    (seq_len, num_batch, -1))
+            compute_in_to_hid = lambda layer_input: layer_input
+        else:
+            compute_in_to_hid = lambda layer_input: helper.get_output(
+                self.input_to_hidden, layer_input, **kwargs)
```

Comment:
Maybe the `input_dot_W` in line 192 should be `input`. We could then get rid of the if else in line 237 + 241

 I would prefer to not use lambda and then only have the if statement.